src/main/kotlin/com/atlassian/performance/tools/awsinfrastructure/api/network/access/SecurityGroupIngressAccessProvider.kt

```diff
@@ -0,0 +1,65 @@
+package com.atlassian.performance.tools.awsinfrastructure.api.network.access
+
+import com.amazonaws.services.ec2.AmazonEC2
+import com.amazonaws.services.ec2.model.*
+
+class SecurityGroupIngressAccessProvider private constructor(
+    private val ec2: AmazonEC2,
+    private val securityGroup: SecurityGroup,
+    private val portRange: IntRange,
+    private val protocol: String
+) : AccessProvider {
+
+    /**
+     * No support for IPv6 yet.
+     */
+    override fun provideAccess(
+        cidr: String
+    ) = provideAccess(
+        baseIpPermissions()
+            .withIpv4Ranges(IpRange().withCidrIp(cidr))
+    )
+
+    override fun provideAccess(
+        securityGroup: SecurityGroup
+    ) = provideAccess(
+        baseIpPermissions()
+            .withUserIdGroupPairs(
+                UserIdGroupPair()
+                    .withUserId(securityGroup.ownerId)
+                    .withGroupId(securityGroup.groupId)
+            )
+    )
+
+    private fun baseIpPermissions() = IpPermission()
+        .withIpProtocol("tcp")
+        .withFromPort(portRange.first)
+        .withToPort(portRange.last)
+        .withIpProtocol(protocol)
+
+    private fun provideAccess(
+        ipPermissions: IpPermission
+    ) = ec2.authorizeSecurityGroupIngress(
+        AuthorizeSecurityGroupIngressRequest()
+            .withGroupId(securityGroup.groupId)
+            .withIpPermissions(ipPermissions)
+    ).sdkHttpMetadata
+        .httpStatusCode / 100 == 2
+
+    class Builder(
+        private val ec2: AmazonEC2,
+        private val securityGroup: SecurityGroup,
+        private val portRange: IntRange
```

Comment:
I can see the only values passed here are single ports across the whole project. Is range needed here?

Reply:
This API was designed to match the underlying mechanism (Security Group Ingress) and it should stay like that, because this models that mechanism. Changing it to single port would be unnecessary limitation.

If we want to achieve simplification of the API we can do another implementation that has only a single port - this one could use the `SecurityGroupIngressAccessProvider`underneath.

I was also experimenting with something different. I believe `DataCenterFormula` should delegate more repsponsibilities outside to constructor injected dependencies. Essentially no dependency objects should be created as part of provisioning itself. We could define "AccessProviderFactory" with API like `fun createAccessProvider(port: Int, securityGroup: SecurityGroup): AccessProvider` (actually that would be `SecurityGroupPortAccessProviderFactory`) and use that instead of calling constructor of specific class. It could provide us an ability to further observe what the formula is doing and API consumers could also override this functionality in different ways. The problem with thts approach is that it fully exposes the internal dependencies of the class - which should be fine if we treat every implementation dependencies as final, however this is not how the code in this repo was made so far.